src/rnnlm/rnnlm-embedding-training.cc

```diff
@@ -175,7 +175,10 @@ void RnnlmEmbeddingTrainer::Train(
   if (config_.l2_regularize > 0.0) {
     BaseFloat l2_term = -2 * config_.l2_regularize;
     if (l2_term != 0.0) {
-      embedding_deriv->AddToRows(l2_term, active_words, embedding_mat_);
+      CuMatrix<BaseFloat> l2_matrix(embedding_deriv->NumRows(),
+                                    embedding_deriv->NumCols(), kUndefined);
```

Comment:
Can you explain what the difference is, or why the old code was wrong?

Reply:
The old code does

a subset of rows of embedding_mat_ += derivs * l2_factor

But it should be

derivs += a subset of rows of embedding_mat * l2_factor


Later derivs will be added to embeeding_mat with learning rate (and possibly momentum) which makes it different.